tests/test_model.py

```diff
@@ -454,18 +454,39 @@ def finalize(self, status):
 
         assert simple_logger_more_saves.saved_count > simple_logger_less_saves.saved_count
 
-    @mock.patch.object(MatrixFactorizationModel, 'calculate_loss')
     def test_terminate_on_nan(self,
-                              calculate_loss_mock,
                               train_val_implicit_sample_data):
-        calculate_loss_mock.return_value = torch.tensor(float('nan')).requires_grad_()
-
         train, val = train_val_implicit_sample_data
         model = MatrixFactorizationModel(train=train, val=val)
-        trainer = CollieMinimalTrainer(model=model, terminate_on_nan=True)
 
-        with pytest.raises(ValueError):
-            trainer.fit(model)
+        with pytest.warns(
+            DeprecationWarning,
+            match=r"``terminate_on_nan`` is deprecated and was replaced with ``detect_anomaly``."
+        ):
+            trainer = CollieMinimalTrainer(model=model, terminate_on_nan=True)
+
+        assert trainer.detect_anomaly
+
+    def test_detect_anomaly(self,
+                            train_val_implicit_sample_data):
+        def bad_loss(*args, **kwargs):
+            return torch.tensor(0.0, requires_grad=True) * torch.tensor(float('nan'))
+
+        train, val = train_val_implicit_sample_data
+        model1 = MatrixFactorizationModel(train=train, val=val, loss=bad_loss)
+        trainer1 = CollieMinimalTrainer(model=model1, detect_anomaly=False, max_epochs=1)
+
+        trainer1.fit(model1)
+
+        model2 = MatrixFactorizationModel(train=train, val=val, loss=bad_loss)
+        trainer2 = CollieMinimalTrainer(model=model2, detect_anomaly=True, max_epochs=1)
+
+        with pytest.raises(RuntimeError, match=r"returned nan values in its 0th output."):
+            with pytest.warns(
+                UserWarning,
+                match=r".*Error detected in.* Traceback of forward call that caused the error.*"
```

Comment:
Is it possible to include the warning message specifically where it mentions a `NaN` loss? I think that might be helpful for future people looking at this test.

Reply:
I was matching the tests from `pytorch_lightning` here https://github.com/PyTorchLightning/pytorch-lightning/blob/e15579a4f32ee3c08318a466583f4a0a8517d654/tests/trainer/test_trainer.py#L2110

I am happy to change it if you feel strongly about it.

Reply:
Eh, it's fine to just keep this as-is then.